lighthouse-logger/index.js

```diff
@@ -118,33 +174,145 @@ class Log {
     return level_ === 'verbose';
   }
 
+  /**
+   * @param {Status} param0
+   * @param {LogAction} level
+   */
   static time({msg, id, args = []}, level = 'log') {
     marky.mark(id);
     Log[level]('status', msg, ...args);
   }
 
+  /**
+   * @param {Status} param0
+   * @param {LogAction} level
+   */
   static timeEnd({msg, id, args = []}, level = 'verbose') {
     Log[level]('statusEnd', msg, ...args);
     marky.stop(id);
   }
 
+  /* eslint-disable no-invalid-this */
+  /**
+   * Decorates a function, calling time/timeEnd before/after calling the original function.
+   * @template T, R
+   * @template {*[]} Args
+   * @param {(this: T, ...args: Args) => R} originalFn
+   * @param {TimeDecorateOpts<T, Args>} opts
+   * @return {(this: T, ...args: Args) => R}
+   */
+  static timeDecorate(originalFn, opts) {
+    /**
+     * @type {(this: T, ...args: Args) => R}
+     */
+    const fn = function timeDecoratedFn(...args) {
+      const timeStartLogLevel = opts.timeStartLogLevel || 'log';
+      const timeEndLogLevel = opts.timeEndLogLevel || 'verbose';
+
+      let msg;
```

Comment:
Not a big deal, but it seems like `msg` and `id` could be turned into functions outside of the `timeDecoratedFn` so we could simplify what's going on inside each invocation a bit, i.e. line 212-234 becomes just `const status = {msg: computeMsg(this, args), id: computeId(this, args)}`

Reply:
done